lib/improver/psychrometric_calculations/psychrometric_calculations.py

```diff
@@ -0,0 +1,411 @@
+
+    Args:
+        temperature : iris.cube.Cube
+            Cube of temperature which will be converted to Celsius
+            prior to calculation
+        rel_humidity : iris.cube.Cube
+            Cube of relative humidity
+        pressure : iris.cube.Cube or float
+            Cube of pressure which will be converted to kilopPascals
+            prior to calculation
+        precision : float
+            degree of precision the user requires for this algorithm
+
+    Returns:
+        result : iris.cube.Cube
+            Cube containing the wet bulb temperature of the air in Kelvin
+
+    Raises:
+        ValueError: If any pressure is 0, as this otherwise causes a divide
+        by zero error later.
+        ValueError: If any relative humidity is 0, as this otherwise causes
+        a divide by zero error later.
+    """
+    # Decouple local data to avoid modifying input cubes
+    temp = temperature.copy()
+    temp.convert_units('celsius')
+    rh = relative_humidity.copy()
+    rh.convert_units(1)
+    if isinstance(pressure, float):
+        press = temperature.copy(data=np.full(temperature.shape, pressure))
+        press.units = Unit(p_units)
+    else:
+        press = pressure.copy()
+    press.convert_units("kPa")
+    # check that lowest pressure and rh values are non-zero
+    if press.data.min() == 0:
+        emsg = ("This Wet Bulb Temperature algorithm is"
+                " only valid for pressures greater than"
+                " Zero. Input cube has\n"
+                "Lowest pressure of:  {}\n")
+        raise ValueError(emsg.format(press.data.min()))
+    if rh.data.min() == 0:
+        emsg = ("This Wet Bulb Temperature algorithm is"
+                " only valid for relative humidities greater than"
+                " Zero. Input cube has\n"
+                "Lowest rh of:  {}\n")
+        raise ValueError(emsg.format(rh.data.min()))
+    # create a numpy array of precision and cube of increment
+    precision = np.full(temp.data.shape, precision)
+    increment = temp.copy(data=np.full(rh.shape, 0.001))
+    hr_normal = humidity_ratio_fm_rh(temp, rh, press)
+    result = temp.copy()
+    hr_new = humidity_ratio_fm_wb(temp, result, press)
+    while ((maths.abs((hr_new - hr_normal) / hr_normal)).data >
+           precision).any():
+        hr_new2 = humidity_ratio_fm_wb(temp, (result - increment), press)
+        dw_dthr = (hr_new - hr_new2) / increment
+        dw_dthr.convert_units('celsius^-1')
+        result.units = Unit(1)
+        dw_dthr.units = Unit(1)
+        result2 = result - (hr_new - hr_normal) / dw_dthr
+        result2.units = Unit('celsius')
+        data = np.where((maths.abs((hr_new - hr_normal) /
+                                   hr_normal)).data > precision,
+                        result2.data,
+                        result.data)
+        result = result.copy(data=data)
+        result.units = Unit('celsius')
+        hr_new = humidity_ratio_fm_wb(temp, result, press)
```

Comment:
This is not very lovely. I've not gone through it in detail, I'd prefer to do that when you reconsider how you've implemented your iterator. Things to consider:
* How much copying is necessary?
* Do the units need to be set at every stage?
* Are there python tools that could help clean this up (scipy.optimize)?

Reply:
I've managed to remove some of the copying and unit changing, but not by any means as much as I'd like.

Inside the while loop there is now only 1 .copy(), which I believe is necessary. There are no unit conversions inside the while loop (the units are forced where they are changed rather than converted.) except for those in the humidity_ratio_fm_wb function.

Part of this ugliness (esp around line ~407) is caused by needing to avoid
`NotYetImplementedError: Cannot use 'subtract' with differing units (celsius & kelvin)`
Sadly Iris does not really have concept of a temperature delta - which is exactly what is needed here.

I had a look at SciPy Optimize, and although there is a Newton-Raphson method, it seems not to be designed for use on arrays, and attempts to do so return an array comparison ambiguous, please use .any() or .all() error.